File: src/services/clustering.ts
/**
 * 新闻聚类服务 (News Clustering)
 * 
 * 借鉴 WorldMonitor 的 Jaccard 相似度聚类：
 * - 词袋交集/并集计算相似度
 * - 同一事件被多个源报道时合并为一个"聚类事件"
 * - 按源 tier 和时新性排序
 * 
 * 核心策略：将碎片化的新闻条目聚合为"宏观事件"，
 * 帮助 AI 在更高层面进行分析。
 */

import { ThreatClassification, aggregateThreats, THREAT_PRIORITY } from './classifier';

export interface NewsItemForClustering {
  id: string;
  title: string;
  url: string;
  content: string;
  source: string;
  sourceTier: number;
  timestamp: string;
  threat: ThreatClassification;
}

export interface ClusteredEvent {
  id: string;
  primaryTitle: string;
  primaryUrl: string;
  primarySource: string;
  sourceCount: number;
  sources: Array<{ name: string; tier: number; title: string; url: string }>;
  allItems: NewsItemForClustering[];
  firstSeen: Date;
  lastUpdated: Date;
  threat: ThreatClassification;
}

// ── 停用词列表（不参与相似度计算）──
const STOP_WORDS = new Set([
  'the', 'a', 'an', 'is', 'are', 'was', 'were', 'be', 'been', 'being',
  'have', 'has', 'had', 'do', 'does', 'did', 'will', 'would', 'could',
  'should', 'may', 'might', 'can', 'shall', 'must', 'to', 'of', 'in',
  'for', 'on', 'with', 'at', 'by', 'from', 'as', 'into', 'through',
  'during', 'before', 'after', 'above', 'below', 'between', 'and',
  'but', 'or', 'nor', 'not', 'so', 'yet', 'both', 'either', 'neither',
  'each', 'every', 'all', 'any', 'few', 'more', 'most', 'other',
  'some', 'such', 'no', 'only', 'own', 'same', 'than', 'too', 'very',
  'just', 'about', 'also', 'then', 'that', 'this', 'these', 'those',
  'it', 'its', 'he', 'she', 'they', 'we', 'you', 'who', 'what',
  'which', 'when', 'where', 'how', 'why', 'if', 'up', 'out', 'over',
  'says', 'said', 'new', 'news', 'report', 'reports', 'according',
]);

/**
 * 分词提取关键 token
 */
function tokenize(text: string): Set<string> {
  const tokens = text
    .toLowerCase()
    .replace(/[^a-z0-9\s'-]/g, ' ')
    .split(/\s+/)
    .filter(t => t.length >= 3 && !STOP_WORDS.has(t));
  return new Set(tokens);
}

/**
 * Jaccard 相似度计算
 */
function jaccardSimilarity(setA: Set<string>, setB: Set<string>): number {
  if (setA.size === 0 || setB.size === 0) return 0;
  let intersection = 0;
  for (const item of setA) {
    if (setB.has(item)) intersection++;
  }
  const union = setA.size + setB.size - intersection;
  return union > 0 ? intersection / union : 0;
}

/**
 * 去除标题末尾的新闻源归属（如 " - Reuters"）
 */
function stripSourceAttribution(title: string): string {
  const idx = title.lastIndexOf(' - ');
  if (idx === -1) return title;
  const after = title.slice(idx + 3).trim();
  if (after.length > 0 && after.length <= 60 && !/[.!?]/.test(after)) {
    return title.slice(0, idx).trim();
  }
  return title;
}

/**
 * 对新闻列表执行聚类
 * 
 * @param items 待聚类的新闻条目
 * @param similarityThreshold Jaccard 相似度阈值（默认 0.3）
 * @returns 聚类后的事件列表，按 threat 优先级和时新性排序
 */
export function clusterNews(
  items: NewsItemForClustering[],
  similarityThreshold: number = 0.3
): ClusteredEvent[] {
  if (items.length === 0) return [];

  // 预计算各标题的 token 集合
  const tokenSets = items.map(item => ({
    item,
    tokens: tokenize(stripSourceAttribution(item.title)),
  }));

  // 并查集分组
  const parent: number[] = tokenSets.map((_, i) => i);
  function find(x: number): number {
    while (parent[x] !== x) {
      parent[x] = parent[parent[x]!]!;
      x = parent[x]!;
    }
    return x;
  }
  function union(a: number, b: number): void {
    const ra = find(a);
    const rb = find(b);
    if (ra !== rb) parent[ra] = rb;
  }

  // O(n²) 两两比较（新闻条目通常 < 500，可接受）
  for (let i = 0; i < tokenSets.length; i++) {
    for (let j = i + 1; j < tokenSets.length; j++) {
      const sim = jaccardSimilarity(tokenSets[i]!.tokens, tokenSets[j]!.tokens);
      if (sim >= similarityThreshold) {
        union(i, j);
      }
    }
  }

  // 按组收集
  const groups = new Map<number, NewsItemForClustering[]>();
  for (let i = 0; i < items.length; i++) {
    const root = find(i);
    const group = groups.get(root) || [];
    group.push(items[i]!);
    groups.set(root, group);
  }

  // 构建聚类事件
  const clusters: ClusteredEvent[] = [];
  for (const members of groups.values()) {
    // 选择 tier 最低（最权威）的作为主标题
    const sorted = [...members].sort((a, b) => {
      if (a.sourceTier !== b.sourceTier) return a.sourceTier - b.sourceTier;
      return new Date(b.timestamp).getTime() - new Date(a.timestamp).getTime();
    });

    const primary = sorted[0]!;
    const dates = members.map(m => new Date(m.timestamp).getTime());
    const firstSeen = new Date(Math.min(...dates));
    const lastUpdated = new Date(Math.max(...dates));

    // 聚合威胁评估
    const threat = aggregateThreats(
      members.map(m => ({ threat: m.threat, tier: m.sourceTier }))
    );

    // 去重源列表
    const sourceMap = new Map<string, { name: string; tier: number; title: string; url: string }>();
    for (const m of sorted) {
      if (!sourceMap.has(m.source)) {
        sourceMap.set(m.source, { name: m.source, tier: m.sourceTier, title: m.title, url: m.url });
      }
    }

    clusters.push({
      id: `cluster-${primary.id}`,
      primaryTitle: primary.title,
      primaryUrl: primary.url,
      primarySource: primary.source,
      sourceCount: sourceMap.size,
      sources: Array.from(sourceMap.values()).slice(0, 5),
      allItems: members,
      firstSeen,
      lastUpdated,
      threat,
    });
  }

  // 按威胁优先级 → 时新性排序
  clusters.sort((a, b) => {
    const pa = THREAT_PRIORITY[a.threat.level];
    const pb = THREAT_PRIORITY[b.threat.level];
    if (pa !== pb) return pb - pa;
    return b.lastUpdated.getTime() - a.lastUpdated.getTime();
  });

  return clusters;
}

/**
 * 生成聚类摘要（用于 AI 上下文）
 */
export function generateClusterContext(clusters: ClusteredEvent[]): string {
  if (clusters.length === 0) return '';

  const lines: string[] = ['[CLUSTERED EVENTS]'];
  const top = clusters.slice(0, 15);

  for (const cluster of top) {
    const sourceInfo = cluster.sourceCount > 1 ? ` (${cluster.sourceCount} sources)` : '';
    const threatTag = cluster.threat.level !== 'info' ? ` [${cluster.threat.level.toUpperCase()}]` : '';
    lines.push(`- ${cluster.primaryTitle}${threatTag}${sourceInfo} via ${cluster.primarySource}`);
  }

  return lines.join('\n');
}

File: src/services/clustering.py
from typing import List, Set, Dict, Any, Optional
from datetime import datetime
from dataclasses import dataclass
from ..config.settings import CONFIG
from ..utils.tokenizer import tokenize, strip_source_attribution
from .classifier import ThreatClassification, aggregate_threats, THREAT_PRIORITY

@dataclass
class NewsItemForClustering:
    id: str
    title: str
    url: str
    content: str
    source: str
    sourceTier: int
    timestamp: str
    threat: ThreatClassification

@dataclass
class ClusteredEvent:
    id: str
    primaryTitle: str
    primaryUrl: str
    primarySource: str
    sourceCount: int
    sources: List[Dict[str, Any]]
    allItems: List[NewsItemForClustering]
    firstSeen: datetime
    lastUpdated: datetime
    threat: ThreatClassification

def jaccard_similarity(set_a: Set[str], set_b: Set[str]) -> float:
    if not set_a or not set_b: return 0.0
    inter = len(set_a & set_b)
    union = len(set_a | set_b)
    return inter / union if union > 0 else 0.0

def cluster_news(
    items: List[NewsItemForClustering],
    similarity_threshold: float = None
) -> List[ClusteredEvent]:
    similarity_threshold = similarity_threshold or CONFIG.cluster_similarity_threshold
    if not items: return []

    token_sets = [
        {"item": item, "tokens": tokenize(strip_source_attribution(item.title))}
        for item in items
    ]

    # 并查集
    n = len(items)
    parent = list(range(n))
    def find(x: int) -> int:
        if parent[x] != x:
            parent[x] = find(parent[x])
        return parent[x]

    def union(a: int, b: int):
        root_a = find(a)
        root_b = find(b)
        if root_a != root_b:
            parent[root_a] = root_b

    for i in range(n):
        for j in range(i + 1, n):
            if jaccard_similarity(token_sets[i]["tokens"], token_sets[j]["tokens"]) >= similarity_threshold:
                union(i, j)

    groups = {}
    for i in range(n):
        root = find(i)
        if root not in groups:
            groups[root] = []
        groups[root].append(items[i])

    clusters = []
    for members in groups.values():
        # 排序：tier 越低越优先，其次时间最新
        sorted_members = sorted(
            members, 
            key=lambda m: (m.sourceTier, -datetime.fromisoformat(m.timestamp.replace('Z', '+00:00')).timestamp())
        )
        
        primary = sorted_members[0]
        timestamps = [datetime.fromisoformat(m.timestamp.replace('Z', '+00:00')) for m in members]
        
        threat = aggregate_threats([{"threat": m.threat, "tier": m.sourceTier} for m in members])
        
        source_map = {}
        for m in sorted_members:
            if m.source not in source_map:
                source_map[m.source] = {
                    "name": m.source, "tier": m.sourceTier, "title": m.title, "url": m.url
                }

        clusters.append(ClusteredEvent(
            id=f"cluster-{primary.id}",
            primaryTitle=primary.title,
            primaryUrl=primary.url,
            primarySource=primary.source,
            sourceCount=len(source_map),
            sources=list(source_map.values())[:5],
            allItems=members,
            firstSeen=min(timestamps),
            lastUpdated=max(timestamps),
            threat=threat
        ))

    # 最终排序：威胁等级最高，其次时间最新
    clusters.sort(
        key=lambda c: (THREAT_PRIORITY[c.threat.level], c.lastUpdated.timestamp()),
        reverse=True
    )

    return clusters

def generate_cluster_context(clusters: List[ClusteredEvent]) -> str:
    if not clusters: return ""

    lines = ["[CLUSTERED EVENTS — Multi-source confirmed macro events]"]
    for i, cluster in enumerate(clusters[:15]):
        idx = i + 1
        source_info = f" ({cluster.sourceCount} independent sources)" if cluster.sourceCount > 1 else ""
        threat_tag = f" [{cluster.threat.level.upper()}]" if cluster.threat.level != 'info' else ""
        lines.append(f"{idx}. {cluster.primaryTitle}{threat_tag}{source_info}")

        if len(cluster.sources) > 1:
            source_names = ", ".join([f"{s['name']}(T{s['tier']})" for s in cluster.sources[:5]])
            lines.append(f"   Sources: {source_names}")

        primary_item = cluster.allItems[0]
        if primary_item.content and len(primary_item.content) > 80:
            excerpt = re.sub(r'\s+', ' ', primary_item.content[:350]).strip()
            lines.append(f"   ▸ {excerpt}")

    return "\n".join(lines)

import re
